### PR TITLE
Add primitive Value Type reference projection support (part 1)

### PR DESCRIPTION
**Pull request by hangshao0**
1. There are 2 ram classes generated from a primitive value type. One is
the L type ram class and the other is the Q type ram class.
2. The L and Q type ram class are companion of each other. They have a
pointer to each other in the J9Class struct.
2. Add new flag J9_FINDCLASS_FLAG_QTYPE to indicate if it is a Q type
that is to be found or generated.
3. Both Q type and L type ram classes are added to the class table. A
new local type TYPE_QTYPE_RAM_CLASS in KeyHashTable.c is added. 
4. The same as existing classes, both L type and Q type ram classes will
be keyed on the class name in the class hash table. However, if it is a
Q type ram class, its hash value will be incremented by 1 (the value of
TYPE_QTYPE_RAM_CLASS)
5. Add J9_HASH_TABLE_QUERY_FLAG_QTYPE to indicate if the hash table
query is for a Q type or not.

issue #14858

Signed-off-by: Hang Shao <hangshao@ca.ibm.com>

### CODE REVIEW

**Comment by hangshao0** [comment]
This change does not cover arrays. I will look at arrays in the next change.

**Comment by gacholio** [comment]
@tajila Can you have the first look? I'm working on a customer issue.

**Comment by tajila** [comment]
Sure Ill take a look

File: runtime/j9vm/j7vmi.c [hunk 1]
```diff
@@ -2627,9 +2627,36 @@ jvmDefineClassHelper(JNIEnv *env, jobject classLoaderObject,
...
 																			NULL,
-																			NULL);
+																			NULL
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+																			, NULL
+#endif /* defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
+																			);
 			/* Done if a class was found or and exception is pending, otherwise try to define the bytes */
 			if ((clazz != NULL) || (currentThread->currentException != NULL)) {
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
```
Comment (tajila) [inline runtime/j9vm/j7vmi.c:2637]: you should check pending exception before doing another class load

File: runtime/j9vm/j7vmi.c [hunk 2]
```diff
@@ -2627,9 +2627,36 @@ jvmDefineClassHelper(JNIEnv *env, jobject classLoaderObject,
...
+																								NULL,
+																								protectionDomain ? *(j9object_t*)protectionDomain : NULL,
+																								NULL,
+																								J9_CP_INDEX_NONE,
+																								LOAD_LOCATION_UNKNOWN,
+																								NULL,
+																								NULL,
+																								clazz);
+					if ((NULL == clazz)
+						&& (NULL == currentThread->currentException)
```
Comment (tajila) [inline runtime/j9vm/j7vmi.c:2654]: should this be `&&` or `||` like the previous check?

File: runtime/jcl/common/jcldefine.c [hunk 3]
```diff
@@ -199,9 +199,36 @@ defineClassCommon(JNIEnv *env, jobject classLoaderObject,
...
+																								NULL,
+																								protectionDomain ? *(j9object_t*)protectionDomain : NULL,
+																								NULL,
+																								J9_CP_INDEX_NONE,
+																								LOAD_LOCATION_UNKNOWN,
+																								NULL,
+																								NULL,
+																								clazz);
+						if ((NULL == clazz)
+							&& (NULL == currentThread->currentException)
```
Comment (tajila) [inline runtime/jcl/common/jcldefine.c:226]: same question here as above

**Review by tajila — COMMENTED** [review]
How do we deal with statics? If one defines a primitive type with a static field, does the static exist in both the L and the Qtype? 

I think that the Qtype can simply reference to the Ltype for things like methods, statics, constantPool, etc. Assuming the behaviour is that both the L and Q are essentially one class, and Im also assuming we create the L type first

**Comment by hangshao0** [comment]
> How do we deal with statics? If one defines a primitive type with a static field, does the static exist in both the L and the Qtype?

Yes, the static exist in both L and Q type. I notice in the class file, statics are currently referenced the same way from both L and Q type. i.e.
 
in class:
```
primitive class PointPrimitive {
        public static int count = 100;
        ...
```

Both `PointPrimitive.ref.count` and `PointPrimitive.count` are referenced via `Field PointPrimitive.count:I` in the class file. I think sharing the statics between L type and Q type or putting them separately does not seem to make much functional difference for now. 

> I think that the Qtype can simply reference to the Ltype for things like methods, statics, constantPool, etc. Assuming the behaviour is that both the L and Q are essentially one class, and Im also assuming we create the L type first

I think we can share the statics, but not sure about the methods and constantPool. We have marcos like `J9_CLASS_FROM_CP()` and `J9_CLASS_FROM_METHOD()`, which will be broken if they are shared between both L and Q type.

File: runtime/util/hshelp.c [hunk 4]
```diff
@@ -2521,6 +2523,9 @@ recreateRAMClasses(J9VMThread * currentThread, J9HashTable * classHashTable, J9H
 		j9object_t protectionDomain = J9VMJAVALANGCLASS_PROTECTIONDOMAIN(currentThread, heapClass);
 		UDATA options;
 		J9Class *hostClass = NULL;
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+		J9Class *companinClass = NULL;
```
[the remark this answers is not in the document]
Comment (hangshao0) [reply runtime/util/hshelp.c]: Fixed.

File: runtime/j9vm/j7vmi.c (hunk 1 above)
Comment (hangshao0) [reply runtime/j9vm/j7vmi.c:2635]: Fxied.

File: runtime/jcl/common/jcldefine.c (hunk 3 above)
Comment (hangshao0) [reply runtime/jcl/common/jcldefine.c]: Fixed.

**Comment by hangshao0** [comment]
Update the code to share ramStatics between L and Q type.

**Comment by hangshao0** [comment]
All existing review comments addressed.

File: runtime/vm/createramclass.cpp [hunk 5]
```diff
@@ -2855,6 +2883,20 @@ internalCreateRAMClassFromROMClassImpl(J9VMThread *vmThread, J9ClassLoader *clas
...
+				/* For shared fragments, set alignedSize and prefixSize to 0 to make internalAllocateRAMClass() ignore them. */
+				allocationRequests[RAM_SUPERCLASSES_FRAGMENT].address = (UDATA *) companionClass->superclasses;
+				allocationRequests[RAM_SUPERCLASSES_FRAGMENT].prefixSize = 0;
+				allocationRequests[RAM_SUPERCLASSES_FRAGMENT].alignedSize = 0;
+				allocationRequests[RAM_INSTANCE_DESCRIPTION_FRAGMENT].address = companionClass->instanceDescription;
+				allocationRequests[RAM_INSTANCE_DESCRIPTION_FRAGMENT].prefixSize = 0;
+				allocationRequests[RAM_INSTANCE_DESCRIPTION_FRAGMENT].alignedSize = 0;
+				allocationRequests[RAM_STATICS_FRAGMENT].address = companionClass->ramStatics;
+				allocationRequests[RAM_STATICS_FRAGMENT].prefixSize = 0;
+				allocationRequests[RAM_STATICS_FRAGMENT].alignedSize = 0;
```
Comment (tajila) [inline runtime/vm/createramclass.cpp:2897]: Are methods ever called on the Ltype? AFAIK the ltype can never be instanciated. If we dont need methods, then we dont need the constant pool, itables, invoke cache, etc.

File: runtime/vm/exceptiondescribe.c [hunk 6]
```diff
@@ -330,7 +330,14 @@ iterateStackTrace(J9VMThread * vmThread, j9object_t* exception, callback_func_t
 					pruneConstructors = FALSE;
 #endif
 					romClass = findROMClassFromPC(vmThread, methodPC, &classLoader);
-					if (NULL != romClass) {
+					if ((NULL != romClass)
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+						&& !J9ROMCLASS_IS_PRIMITIVE_VALUE_TYPE(romClass)
```
Comment (tajila) [inline runtime/vm/exceptiondescribe.c:335]: could we peek but then always use the qtype if one exists?

File: runtime/vm/visible.c [hunk 7]
```diff
@@ -118,7 +118,12 @@ checkVisibility(J9VMThread *currentThread, J9Class* sourceClass, J9Class* destCl
 #endif /* JAVA_SPEC_VERSION >= 11 */
 		} else if (modifiers & J9AccPrivate) {
 			/* Private */
-			if (sourceClass != destClass) {
+			BOOLEAN sameClass = (sourceClass == destClass);
```
Comment (tajila) [inline runtime/vm/visible.c:121]: nitpick: may be more clear this way

`BOOLEAN sameClass = (sourceClass == destClass) || J9_ARE_J9CLASSES_COMPANIONS(sourceClass, destClass)`

File: runtime/vm/ClassInitialization.cpp [hunk 8]
```diff
@@ -418,7 +420,7 @@ classInitStateMachine(J9VMThread *currentThread, J9Class *clazz, J9ClassInitStat
 							U_8 *signatureChars = J9UTF8_DATA(signature);
 
 							PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, initializationLock);
-							J9Class *valueClass = internalFindClassUTF8(currentThread, signatureChars + 1, J9UTF8_LENGTH(signature) - 2, clazz->classLoader, J9_FINDCLASS_FLAG_THROW_ON_FAIL);
+							J9Class *valueClass = internalFindClassUTF8(currentThread, signatureChars, J9UTF8_LENGTH(signature), clazz->classLoader, J9_FINDCLASS_FLAG_THROW_ON_FAIL);
```
Comment (tajila) [inline runtime/vm/ClassInitialization.cpp:423]: once the Qtype is initialized, is there any need to initialize the ltype? If it has no members or state, perhaps it should just be automatically initialized once the qtype is initialized. In other words can they share intializations state

File: runtime/util/superclass.c [hunk 9]
```diff
@@ -29,11 +29,20 @@
 UDATA 
 isSameOrSuperClassOf(J9Class *superClass, J9Class *baseClass)
 {
-	UDATA superClassDepth = J9RAMCLASS_DEPTH(superClass);
 	
-	return ((baseClass == superClass)
-					|| ((J9RAMCLASS_DEPTH(baseClass) > superClassDepth)
-					   && ((J9RAMCLASS_SUPERCLASSES(baseClass)[superClassDepth]) == superClass)));
+	UDATA ret = TRUE;
```
Comment (tajila) [inline runtime/util/superclass.c:33]: Perhaps we need to add a `VMHelper::isSameClass(clazz1, clazz2)` that does the companion test so simplyfy the code

File: runtime/oti/vm_api.h [hunk 10]
```diff
@@ -485,7 +485,11 @@ popLoadingOrLinkingStack(J9VMThread *vmThread, J9StackElement **stack, J9Pool *s
 J9Class *
 internalCreateRAMClassFromROMClass(J9VMThread *vmThread, J9ClassLoader *classLoader, J9ROMClass *romClass,
 	UDATA options, J9Class* elementClass, j9object_t protectionDomain, J9ROMMethod ** methodRemapArray,
-	IDATA entryIndex, I_32 locationType, J9Class *classBeingRedefined, J9Class *hostClass);
+	IDATA entryIndex, I_32 locationType, J9Class *classBeingRedefined, J9Class *hostClass
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+	, J9Class *companionClass
```
Comment (tajila) [inline runtime/oti/vm_api.h:490]: nitpick, the comma can be on `hostClass, `

File: runtime/cfdumper/stubs.c [hunk 11]
```diff
@@ -57,7 +57,11 @@ void freeMemorySegment (J9JavaVM *javaVM, J9MemorySegment *segment, BOOLEAN free
 J9Class *   
 internalCreateRAMClassFromROMClass(J9VMThread *vmThread, J9ClassLoader *classLoader, J9ROMClass *romClass,
 	UDATA options, J9Class* elementClass, j9object_t protectionDomain, J9ROMMethod ** methodRemapArray,
-	IDATA entryIndex, I_32 locationType, J9Class *classBeingRedefined, J9Class *hostClass)
+	IDATA entryIndex, I_32 locationType, J9Class *classBeingRedefined, J9Class *hostClass
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+	, J9Class *companionClass
```
Comment (tajila) [inline runtime/cfdumper/stubs.c:62]: same here

File: runtime/oti/VMHelpers.hpp [hunk 12]
```diff
@@ -538,13 +538,18 @@ class VM_VMHelpers
 	isSameOrSuperclass(J9Class *superClass, J9Class *subClass)
 	{
 		bool isSubclass = true;
-		if (subClass != superClass) {
+		if ((subClass != superClass)
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+			&& !J9_ARE_J9CLASSES_COMPANIONS(superClass, subClass)
```
[the remark this answers is not in the document]
Comment (hangshao0) [reply runtime/oti/VMHelpers.hpp:543]: I can do assignment between Point.ref and Point, e.g.

```
Point.ref rp = new Point(0,0);
Point p = rp;
```

File: runtime/vm/createramclass.cpp (hunk 5 above)
Comment (hangshao0) [reply runtime/vm/createramclass.cpp:2897]: > Are methods ever called on the Ltype? AFAIK the ltype can never be instanciated.

For the methods, I don't see they are called on Ltype now from bytecodes generated from javac. But I see the following in the [JEP](https://openjdk.org/jeps/401):

`The aconst_init instruction may refer to either a primitive type or a reference type. This determines whether a primitive value or a value object is produced.`

File: runtime/vm/visible.c (hunk 7 above)
Comment (hangshao0) [reply runtime/vm/visible.c]: Fixed. Using the new macro J9_ARE_J9CLASSES_SAME().

File: runtime/oti/vm_api.h (hunk 10 above)
Comment (hangshao0) [reply runtime/oti/vm_api.h:490]: Won't compile in non-Valhalla build if comma moved on `hostClass, `

File: runtime/cfdumper/stubs.c (hunk 11 above)
Comment (hangshao0) [reply runtime/cfdumper/stubs.c:62]: Won't compile in non-Valhalla build if comma moved on hostClass,

File: runtime/vm/exceptiondescribe.c (hunk 6 above)
Comment (hangshao0) [reply runtime/vm/exceptiondescribe.c:335]: I believe it depends on whether Ltype can be instanciated, whether we can run methods on Ltype.

File: runtime/vm/createramclass.cpp (hunk 5 above)
Comment (tajila) [reply runtime/vm/createramclass.cpp:2897]: I believe the `or a reference type.` was writtin in the context of Value classes which do not have a primitive type variant. Essentially, there can only be one instance type.

Comment (hangshao0) [reply runtime/vm/createramclass.cpp:2897]: Ah, I see.

File: runtime/vm/ClassInitialization.cpp (hunk 8 above)
Comment (hangshao0) [reply runtime/vm/ClassInitialization.cpp:428]: Update the code to share the intializations state between L type and Q type.

**Comment by hangshao0** [comment]
I noticed the Valhalla build is currently broken by missing classes in the JCL:

```
import jdk.internal.vm.ContinuationScope;
                      ^
  symbol:   class ContinuationScope
  location: package jdk.internal.vm
/root/CCM/JdkNext/openj9vt/openj9-openjdk-jdk.valuetypes/build/linux-x86_64-server-release/support/j9jcl/java.base/share/classes/java/lang/StackWalker.java:237: error: cannot find symbol
        static StackWalker newInstance(Set<Option> options, ExtendedOption extendedOption, ContinuationScope contScope) {
```

Talked to @JasonFengJ9, there seem to be some JCL merge conflicts that some changes did not go into https://github.com/ibmruntimes/openj9-openjdk-jdk.valuetypes.

**Comment by JasonFengJ9** [comment]
Updated [openj9-openjdk-jdk.valuetypes](https://github.com/ibmruntimes/openj9-openjdk-jdk.valuetypes)
@hangshao0 you might give another try.

**Comment by hangshao0** [comment]
> you might give another try.

Things are working now. Thanks @JasonFengJ9

File: runtime/vm/createramclass.cpp (hunk 5 above)
Comment (hangshao0) [reply runtime/vm/createramclass.cpp:2897]: > Are methods ever called on the Ltype? AFAIK the ltype can never be instanciated. If we dont need methods, then we dont need the constant pool, itables, invoke cache, etc.

Do you want to set these fields to NULL on the Ltype ? I see we have a lot of places that use them directly without NULL check, for example:

https://github.com/eclipse-openj9/openj9/blob/6a2aac87ec442d1d96863265761996c9764b19f3/runtime/codert_vm/jswalk.c#L1005-L1008

https://github.com/eclipse-openj9/openj9/blob/6a2aac87ec442d1d96863265761996c9764b19f3/runtime/rasdump/heapdump.cpp#L717

https://github.com/eclipse-openj9/openj9/blob/6a2aac87ec442d1d96863265761996c9764b19f3/runtime/vmchk/checkclasses.c#L198-L199

https://github.com/eclipse-openj9/openj9/blob/6a2aac87ec442d1d96863265761996c9764b19f3/runtime/jvmti/jvmtiExtensionMechanism.c#L2556-L2562

**Comment by gacholio** [comment]
It seems to me this is being implemented before we understand the requirements.

Can you please describe exactly the function of the Q and L types? What is the result of `Something.class`? Are the statics supposed to be shared or should they only be referenced from one or the other of the classes?

I'm concerned that sharing statics may lead to Unsafe not working properly - the static address is stored as an offset from the ramStatics of the Field instance, so using the wrong class in the Unsafe call will result in a crash/corruption.

**Comment by gacholio** [comment]
> Can you please describe exactly the function of the Q and L types?

I guess I'm really interested in the circumstances where the Q and L types would be used.

**Comment by hangshao0** [comment]
> Can you please describe exactly the function of the Q and L types?  

There are 2 types, reference type (L type) and value type (Q type) generated from the same VT class file. The L type is used if user want to allow NULL or atomic access. It won't be flattened in its container. Q type does not allow NULL, it does not guarantee atomic access and it can be flattened. From the [JEP](https://openjdk.org/jeps/401), the relationship between the Q type (`Point`) and L type (`Point.ref`) is similar to the traditional relationship between the types `int` and `Integer`. However, `Point` and `Point.ref` both correspond to the same class declaration; the values of both types are instances of a single `Point` class.

> What is the result of Something.class?

From the current RI implementation, `Something.class` is the L type.  There are 2 new fields in Class.java:
```
private transient Class<T> primaryType;
private transient Class<T> secondaryType;
```
They are set only when the class is value type. `primaryType` is set to the class representing the L type and `secondaryType` is set to the class representing the Q type.

> Are the statics supposed to be shared or should they only be referenced from one or the other of the classes?

Currently the statics can only be referenced from the Q type, not the L type. 

Not sue if you have anything to add @tajila

**Comment by gacholio** [comment]
> Currently the statics can only be referenced from the Q type, not the L type.

This is somewhat inconsistent with Something.class being the L type - I would expect everything to be accessed via the user-facing type. If that's not the case, there's going to be a lot of unnecessary indirection to keep flipping between the types (perhaps need to do j9class->heapClass->primaryType->j9class to get to the right place).

**Comment by hangshao0** [comment]
There is inconsistence between the RI implementation and the JEP documentation, not sure which one they are going to update.
The JEP says "a `CONSTANT_Class` using the plain name of a primitive class represents the class's reference type", RI JCL sets primaryType to the reference type, but the RI JVM treats `CONSTANT_Class` of the plain name as Q type instead of reference type, therefore statics are referenced from the Q type.

**Comment by gacholio** [comment]
Are all classrefs plain types now or are they a mixture of Q, L and plain?

**Comment by hangshao0** [comment]
They are mixture of Q, L and plain.

**Comment by hangshao0** [comment]
> Are there java.lang.Class instances for both the Q and L types?

Yes, one java.lang.Class instance for Q type and one java.lang.Class instance for L type.

**Comment by hangshao0** [comment]
> What is the behaviour of getClass()? If used on a Q instance do you get the Q class back, or is the L type always the user-facing class? 

`getClass()` returns the L type class on a Q instance. But there are public getters in `Class.java` that return the `primaryType` and `secondaryType`. so the Q type class can be exposed to the users as well.


> Thinking of the Field class - variables will have different offsets in the Q and L instances. Similarly for Method - are you expected to be able to use the Q and L classes to invoke (presumably identical) methods?

According to a previous [comment](https://github.com/eclipse-openj9/openj9/pull/15070#discussion_r896815466) from Tobi, L type class cannot be instanciated, we could only have instance of the Q type class.

For a class like:
```
primitive class PointPrimitive {
        public static int count = 100;
        ...
```

Both `PointPrimitive.ref.count` and `PointPrimitive.count` are compiled to `Field PointPrimitive.count:I` by javac. If plain class name is treated as Q type, they are both accessed from the Q type. It is the same for static methods.

**Comment by gacholio** [comment]
It's not at all confusing for the Q class to be the actual type and the L class to be what people see. :)

I guess we just need to know what happens if people use the Q class in various APIs like reflect.

As an aside does `Class.newInstance()` work on Q types (presumably via the L class)?

Are we linking the RAM classes so we can avoid transitioning through the objects to get from Q to L and back?

**Comment by hangshao0** [comment]
> As an aside does Class.newInstance() work on Q types (presumably via the L class)?

No, it does not work. It triggers `java.lang.InstantiationException`.

> Are we linking the RAM classes so we can avoid transitioning through the objects to get from Q to L and back?

Yes, there is a new filed `J9Class* companionClass` add in `J9Class` struct. L and Q type J9Classes point to each other via this field.

**Comment by gacholio** [comment]
Wondering if it would make sense to add another field like `primaryClass` which would be the same on both Q and L, so a single fetch with no test and branch can be used to get the "correct" class (presumably, Q is almost always the interesting version).

**Comment by gacholio** [comment]
Given everything said above, I'd rather not share the statics - makes more sense to keep them on the Q class and make sure we're always using the Q when we access them.

**Comment by gacholio** [comment]
If we do want to add a new field, maybe `companionClass` should be renamed - we could explicitly have Q and L class pointers initialized appropriately (and for a normal class, they both point to itself). I'd rather avoid test and branch where possible.

**Comment by hangshao0** [comment]
Do we want to set the statics (constant pool, itables, invoke cache, etc.) on the L type J9Class or leave them to NULL ? There are a lot of places they are used without NULL check as in https://github.com/eclipse-openj9/openj9/pull/15070#discussion_r899248833

**Comment by gacholio** [comment]
Where possible, I'd like to see them left NULL. Hopefully most things will be covered by associated counters, so the NULL pointer may be read but will never actually be dereferenced.

**Comment by hangshao0** [comment]
Noticed some counters are on the romClass (like `romMethodCount` for methods), which is not 0.

**Comment by gacholio** [comment]
Then we need a different solution - duplicating the pointers will make it very easy for these things to be walked twice by the GC (which may or may not be an issue but it's a bad idea in general) or other things like DDR-based tools.

**Comment by gacholio** [comment]
The easiest solution would be to just have dummy (empty) entries in the L class - it's a waste of space, but it's simple.

**Comment by hangshao0** [comment]
This can be reviewed again.

**Comment by hangshao0** [comment]
I've resolved the merge conflict and it can be reviewed again.

**Comment by hangshao0** [comment]
Any review comments ? @tajila @gacholio

**Comment by gacholio** [comment]
I'll try to have a look tomorrow.

**Comment by hangshao0** [comment]
I'm closing this PR as it is a bit out-of-date. I am opening a new PR https://github.com/eclipse-openj9/openj9/pull/17060 to replace this one.

**Pull request closed without merging**